### PR TITLE
Feat/add welcome tutorial notebook for new users

### PR DESCRIPTION
**Pull request by daukadolt**
## Description

Update default tutorial:
1. Use tutorial from [notebook_tutorial branch](https://github.com/topoteretes/cognee/blob/notebook_tutorial/notebooks/tutorial.ipynb), specifically - it's .zip version with all necessary data files
2. Use Jupyter Notebook `Notebook` abstractions to read, and map `ipynb` into our Notebook model
3. Dynamically update starter notebook code blocks that reference starter data files, and swap them with local paths to downloaded copies
4. Test coverage

| Before | After (storage backend = local) | After (s3) |
|--------|---------------------------------|------------|
| Screenshot 2025-09-17 at 01 00 58 | Screenshot 2025-09-18 at 13 01 57 | Screenshot 2025-09-18 at 12 56 08 |

### S3 Demo  
https://github.com/user-attachments/assets/bd46eec9-ef77-4f69-9ef0-e7d1612ff9b3

---

### Local FS Demo  
https://github.com/user-attachments/assets/8251cea0-81b3-4cac-a968-9576c358f334

## Type of Change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update
- [ ] Code refactoring
- [ ] Performance improvement
- [ ] Other (please specify):

## Changes Made

- 
- 
- 

## Pre-submission Checklist

- [ ] **I have tested my changes thoroughly before submitting this PR**
- [ ] **This PR contains minimal changes necessary to address the issue/feature**
- [ ] My code follows the project's coding standards and style guidelines
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if applicable)
- [ ] All new and existing tests pass
- [ ] I have searched existing PRs to ensure this change hasn't been submitted already
- [ ] I have linked any relevant issues in the description
- [ ] My commits have clear and descriptive messages

## DCO Affirmation
I affirm that all code in every commit of this pull request conforms to the terms of the Topoteretes Developer Certificate of Origin.

### CODE REVIEW

File: cognee/modules/users/methods/create_user.py [hunk 1]
```diff
@@ -14,6 +14,43 @@
 from typing import Optional
 
 
+async def _create_tutorial_notebook(
```
Comment (borisarzentar) [inline cognee/modules/users/methods/create_user.py]: Let's move this to notebooks module. It may be reused for other purposes as well.

File: cognee/tests/unit/modules/users/test_tutorial_notebook_creation.py [hunk 2]
```diff
@@ -0,0 +1,450 @@
...
+        invalid_json = "{ invalid json content"
+
+        from nbformat.reader import NotJSONError
+
+        with pytest.raises(NotJSONError):
+            Notebook.from_ipynb_string(notebook_content=invalid_json, owner_id=user_id)
+
+    @pytest.mark.asyncio
+    @patch.object(Notebook, "from_ipynb_zip_url")
+    async def test_create_tutorial_notebook_success(self, mock_from_zip_url, mock_session):
```
Comment (borisarzentar) [inline cognee/tests/unit/modules/users/test_tutorial_notebook_creation.py]: Let's try to find a way to test this without mocking internals. We know inputs and expected outputs of the function it is calling, nothing else.
We use sqlite, so it is quite easy to setup cognee and try it that way. You can look in other tests, we usually set data and system directories, call setup(), run test, prune() at the end to clean up everything.

File: cognee/tests/unit/modules/users/test_tutorial_notebook_creation.py [hunk 3]
```diff
@@ -0,0 +1,450 @@
...
+    async def test_create_tutorial_notebook_error_propagated(self, mock_from_zip_url, mock_session):
+        """Test that errors are propagated when zip fetch fails."""
+        user_id = uuid4()
+        mock_from_zip_url.side_effect = Exception("Network error")
+
+        # Should raise the exception (not catch it)
+        with pytest.raises(Exception, match="Network error"):
+            await _create_tutorial_notebook(user_id, mock_session)
+
+        # Verify error handling path was taken
```
Comment (borisarzentar) [inline cognee/tests/unit/modules/users/test_tutorial_notebook_creation.py:157]: Same here, we should find a way to verify this without mocking internals.

File: cognee/tests/unit/modules/users/test_tutorial_notebook_creation.py [hunk 4]
```diff
@@ -0,0 +1,450 @@
...
+    def test_generate_cell_name_with_markdown_header(self):
+        """Test cell name generation from markdown headers."""
+        from nbformat.notebooknode import NotebookNode
+
+        # Mock a markdown cell with header
+        mock_cell = NotebookNode(
+            {"cell_type": "markdown", "source": "# This is a Header\n\nSome content here."}
+        )
+
+        result = Notebook._generate_cell_name(mock_cell)
```
Comment (borisarzentar) [inline cognee/tests/unit/modules/users/test_tutorial_notebook_creation.py]: Isn't this already tested previously in the import test, we have couple of tests testing cell name?

**Review by borisarzentar — COMMENTED** [review]
It is a nice implementation, and we can reuse it for import notebook feature. Just to reorganize it a bit.
I feel that a lot of tests are written by AI, some are repeating and a lot of tests exposing internals. Let's find a way to reduce the amount of tests and use minimal amount of mocks.

File: cognee/tests/unit/modules/users/test_tutorial_notebook_creation.py (hunk 3 above)
Comment (daukadolt) [reply cognee/tests/unit/modules/users/test_tutorial_notebook_creation.py:157]: I removed all internal mockings, this is the only one I left as it's to validate that `from_ipynb_zip_url` doesn't swallow or silence exceptions

File: cognee/tests/unit/modules/users/test_tutorial_notebook_creation.py (hunk 4 above)
Comment (daukadolt) [reply cognee/tests/unit/modules/users/test_tutorial_notebook_creation.py]: these are to cover edge cases for naming, but since these are not that critical (as we don't display markdown cell names), I'll keep just one for code blocks

**Comment by daukadolt** [comment]
rebased and resolved merge conflict

File: pyproject.toml [hunk 5]
```diff
@@ -39,13 +39,15 @@ dependencies = [
     "langfuse>=2.32.0,<3",
     "filetype>=1.2.0,<2.0.0",
     "aiohttp>=3.11.14,<4.0.0",
+    "requests>=2.25.0,<3.0.0",
```
Comment (borisarzentar) [inline pyproject.toml]: We shouldn't introduce another dependency for sending requests, we have already `aiohttp`. Could we use that one?

File: cognee/shared/cache.py [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+from pathlib import Path
+from typing import Optional, Tuple
+import requests
+import logging
+
+logger = logging.getLogger(__name__)
+
+
+def delete_cache():
+    """Delete the Cognee cache directory."""
```
Comment (borisarzentar) [inline cognee/shared/cache.py]: Will this cache and this way of getting notebook work in saas as well?

Comment (daukadolt) [reply cognee/shared/cache.py]: Good flag. I’m working on the S3-compatible path handling and will update once that’s in place.

Comment (daukadolt) [reply cognee/shared/cache.py]: On another note - since we’ll be running locally first, should we just proceed with the current version?

The S3 cache would only come into play if notebooks were running on SaaS - and I’m not even sure that’s something we’d actually want to happen.

Comment (Vasilije1990) [reply cognee/shared/cache.py]: @daukadolt the whole point it to have it on SaaS so people have an onboarding experience

**Comment by daukadolt** [comment]
## File Replacements

### S3 Demo  
https://github.com/user-attachments/assets/bd46eec9-ef77-4f69-9ef0-e7d1612ff9b3

---

### Local FS Demo  
https://github.com/user-attachments/assets/8251cea0-81b3-4cac-a968-9576c358f334

File: pyproject.toml (hunk 5 above)
Comment (daukadolt) [reply pyproject.toml]: Replacing requests with aiohttp in this PR turned out a bit tricky due to SSL issues.
I’ll handle the async migration in a follow-up, so I think it’s fine not to block this PR on that.

File: cognee/shared/cache.py [hunk 7]
```diff
@@ -0,0 +1,261 @@
...
+    def __init__(self, cache_subdir: str = "cache"):
+        """
+        Initialize the cache manager.
+
+        Args:
+            cache_subdir: Subdirectory name within the system root for caching
+        """
+        self.base_config = get_base_config()
+        self.cache_base_path = f"{cache_subdir}"
+        self.storage_manager = get_file_storage(self.base_config.system_root_directory)
```
Comment (borisarzentar) [inline cognee/shared/cache.py]: You can also call `get_file_storage()` with the `cache` subdirectory, so you know that you always work in that subdirectory. You are operating in the `system_root_directory`, which contains users' databases, so be careful.

File: cognee/shared/cache.py [hunk 8]
```diff
@@ -0,0 +1,261 @@
...
+                cached_version = (await asyncio.to_thread(f.read)).strip()
+                return cached_version == version_or_hash
+        except Exception as e:
+            logger.debug(f"Error checking cache validity: {e}")
+            return False
+
+    async def _clear_cache(self, cache_dir: str) -> None:
+        """Clear a cache directory."""
+        try:
+            await self.storage_manager.remove_all(cache_dir)
```
Comment (borisarzentar) [inline cognee/shared/cache.py:104]: This is very dangerous when storage_manager operates on `system_root_directory`.

File: cognee/modules/notebooks/models/Notebook.py [hunk 9]
```diff
@@ -51,3 +62,208 @@ class Notebook(Base):
...
+        import re
+        from cognee.base_config import get_base_config
+        from cognee.infrastructure.files.storage.get_file_storage import get_file_storage
+        from cognee.shared.logging_utils import get_logger
+
+        logger = get_logger()
+
+        # Get storage manager for the cache directory
+        base_config = get_base_config()
+        storage_manager = get_file_storage(base_config.system_root_directory)
```
Comment (borisarzentar) [inline cognee/modules/notebooks/models/Notebook.py]: Also here, better to set it already in `cache` or `cache/data` subdirectory, to reduce the chances of accidental deletion of users' databases.

File: cognee/shared/cache.py (hunk 7 above)
Comment (daukadolt) [reply cognee/shared/cache.py]: My initial thought was to keep the cache folder there since I considered it part of the system folder.

I understand the concern though, let me update it
